modules/service/src/main/scala/database/Database.scala

```diff
@@ -68,215 +69,262 @@ trait Database[F[_]] {
 
 object Database extends Codecs {
 
-  def fromSession[F[_]: Sync](s: Session[F]): Database[F] =
+  def fromSession[F[_]: Sync: Trace](s: Session[F]): Database[F] =
     new Database[F] {
 
       def canonicalizeServiceUser(serviceName: String): F[ServiceUser] =
-        s.prepare(CanonicalizeServiceUser).use(_.unique(serviceName))
+        Trace[F].span("canonicalizeServiceUser") {
+          s.prepare(CanonicalizeServiceUser).use(_.unique(serviceName))
+        }
 
       def createApiKey(roleId: StandardRole.Id): F[ApiKey] =
-        s.prepare(CreateApiKey).use(_.unique(roleId))
+        Trace[F].span("createApiKey") {
+          s.prepare(CreateApiKey).use(_.unique(roleId))
+        }
 
       def deleteApiKey(keyId: PosLong): F[Unit] =
-        s.prepare(DeleteApiKey).use(_.execute(keyId)).void
+        Trace[F].span("deleteApiKey") {
+          s.prepare(DeleteApiKey).use(_.execute(keyId)).void
+        }
 
       def createGuestUser: F[GuestUser] =
-        s.unique(InsertGuestUser)
+        Trace[F].span("createGuestUser") {
+          s.unique(InsertGuestUser)
+        }
 
       def createGuestUserSessionToken(guestUser: GuestUser): F[SessionToken] =
-        s.prepare(InsertGuestUserSessionToken).use(_.unique(guestUser.id))
+        Trace[F].span("createGuestUserSessionToken") {
+          s.prepare(InsertGuestUserSessionToken).use(_.unique(guestUser.id))
+        }
 
       def createGuestUserAndSessionToken: F[(GuestUser, SessionToken)] =
-        s.transaction.use { _ =>
-          createGuestUser.mproduct(createGuestUserSessionToken) // 🔥
+        Trace[F].span("createGuestUserAndSessionToken") {
+          s.transaction.use { _ =>
+            createGuestUser.mproduct(createGuestUserSessionToken) // 🔥
```

Comment:
Just out of curiosity, what is the significance of the fire emoji comment?

Reply:
Functor has `fproduct : F[A] => (A => B) => F[(A, B)]` which is pretty well known. Lesser know is monad's `mproduct: M[A] => (A => M[B]) => M[(A, B)]` so I was excited I had a change to use it.

Reply:
Nice. 😄